### PR TITLE
castopod: 1.6.4 -> 1.6.5, some fixes

### PR DESCRIPTION
**Pull request by AleXoundOS**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Izorkin — APPROVED** [review]

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 260518` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>castopod</li>
  </ul>
</details>

File: nixos/modules/services/audio/castopod.nix
Comment (balsoft) [inline nixos/modules/services/audio/castopod.nix:252]: Please make this configurable

Comment (AleXoundOS) [reply nixos/modules/services/audio/castopod.nix]: Done.

**Comment by sinavir** [comment]
Hello, really nice work !

There is a bug in the module at line 206. The parenthesis is closed and never opened

**Comment by sinavir** [comment]
Also the exemple in the manual have nginx options at the wrong place

**Comment by AleXoundOS** [comment]
@sinavir, thank you very much for pointing out mistakes!

NixOS castopod package&service is a joint effort of several people.

---

As for the nginx options, I find some examples of NixOS services with such pattern:
```nix
    nginx = mkOption {
      type = types.submodule (
        recursiveUpdate
          (import ../web-servers/nginx/vhost-options.nix { inherit config lib; }) {}
      );
```
But I'm not sure it's the right way, because it bloats the package and manual with lots of nginx options (inherited from nginx). `castopod.md` assumed exactly this implementation (but it's not implemented in code). What do you think?

**Comment by sinavir** [comment]
> @sinavir, thank you very much for pointing out mistakes!
> 
> NixOS castopod package&service is a joint effort of several people.
> 
> As for the nginx options, I find some examples of NixOS services with such pattern:
> 
> ```nix
>     nginx = mkOption {
>       type = types.submodule (
>         recursiveUpdate
>           (import ../web-servers/nginx/vhost-options.nix { inherit config lib; }) {}
>       );
> ```
> 
> But I'm not sure it's the right way, because it bloats the package and manual with lots of nginx options (inherited from nginx). Example in `castopod.md` assumed exactly this implementation (but it's not implemented in code). What do you think?

I do not know the nixpkgs standard on this subject. I was just pointing that the doc wasn't sticking to the implementation but I see that it is not the case anymore.

**Comment by AleXoundOS** [comment]
After rebasing, selenium driver is unable to initialize, hence NixOS tests cannot run:
```
vm-test-run-castopod> client # Traceback (most recent call last):
vm-test-run-castopod> client #   File "/nix/store/jdm5chqkmh3z0wbip7r165fas3bfg1y4-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/driver_finder.py", line 38, in get_path
vm-test-run-castopod> client #     path = SeleniumManager().driver_location(options) if path is None else path
vm-test-run-castopod> client #            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/jdm5chqkmh3z0wbip7r165fas3bfg1y4-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/selenium_manager.py", line 78, in driver_location
vm-test-run-castopod> client #     args = [str(self.get_binary()), "--browser", browser]
vm-test-run-castopod> client #                 ^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/jdm5chqkmh3z0wbip7r165fas3bfg1y4-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/selenium_manager.py", line 57, in get_binary
vm-test-run-castopod> client #     if not path.is_file() and os.environ["CONDA_PREFIX"]:
vm-test-run-castopod> client #                               ~~~~~~~~~~^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "<frozen os>", line 679, in __getitem__
vm-test-run-castopod> client # KeyError: 'CONDA_PREFIX'
vm-test-run-castopod> client # 
vm-test-run-castopod> client # The above exception was the direct cause of the following exception:
vm-test-run-castopod> client # 
vm-test-run-castopod> client # Traceback (most recent call last):
vm-test-run-castopod> client #   File "/run/current-system/sw/bin/browser-test", line 10, in <module>
vm-test-run-castopod> client #     driver = Firefox(options=options)
vm-test-run-castopod> client #              ^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/jdm5chqkmh3z0wbip7r165fas3bfg1y4-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/firefox/webdriver.py", line 59, in __init__
vm-test-run-castopod> client #     self.service.path = DriverFinder.get_path(self.service, options)
vm-test-run-castopod> client #                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/jdm5chqkmh3z0wbip7r165fas3bfg1y4-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/driver_finder.py", line 41, in get_path
vm-test-run-castopod> client #     raise NoSuchDriverException(msg) from err
vm-test-run-castopod> client # selenium.common.exceptions.NoSuchDriverException: Message: Unable to obtain driver for firefox using Selenium Manager.; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location
```

[2023-09-15 selenium has been updated in nixpkgs](https://github.com/NixOS/nixpkgs/commit/612f8e91417ec1985cc865b7c2b42819f7a0ece9), what supposedly broke its compatibility with the gecko driver.

**Comment by AleXoundOS** [comment]
@mweinelt, hi! Do you know how to fix selenium for running the NixOS test?

**Comment by mweinelt** [comment]
The first traceback is a regression in 4.12.0. I've provided a fix upstream at https://github.com/SeleniumHQ/selenium/pull/13071.

**Comment by AleXoundOS** [comment]
@mweinelt, thank you for getting into this!

However, it seems that the problem is more general, as now selenium is tied with [selenium-manager](https://www.selenium.dev/documentation/selenium_manager/) (which Nix obviously doesn't use), which downloads browsers/drivers. [Changes in selenium upstream](https://github.com/SeleniumHQ/selenium/commit/6176d7ec65#diff-1e444af73ad0e1c989ee6c2b0e03dd257071fe4a165e450a23bd86bea7d26323L25) have made [`substituteInPlace` call](https://github.com/NixOS/nixpkgs/blob/66f99272aa70bb8e085dd6b94072796bdb83ce0b/pkgs/development/python-modules/selenium/default.nix#L30) obsolete.


I've tried your downstream fix:
```
vm-test-run-castopod> client #   File "/nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/driver_finder.py", line 38, in get_path
vm-test-run-castopod> client #     path = SeleniumManager().driver_location(options) if path is None else path
vm-test-run-castopod> client #            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/selenium_manager.py", line 78, in driver_location
vm-test-run-castopod> client #     args = [str(self.get_binary()), "--browser", browser]
vm-test-run-castopod> client #                 ^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/selenium_manager.py", line 62, in get_binary
vm-test-run-castopod> client #     raise WebDriverException(f"Unable to obtain working Selenium Manager binary; {path}")
vm-test-run-castopod> client # selenium.common.exceptions.WebDriverException: Message: Unable to obtain working Selenium Manager binary; /nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/linux/selenium-manager
vm-test-run-castopod> client # 
vm-test-run-castopod> client # 
vm-test-run-castopod> client # The above exception was the direct cause of the following exception:
vm-test-run-castopod> client # 
vm-test-run-castopod> client # Traceback (most recent call last):
vm-test-run-castopod> client #   File "/run/current-system/sw/bin/browser-test", line 10, in <module>
vm-test-run-castopod> client #     driver = Firefox(options=options)
vm-test-run-castopod> client #              ^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/firefox/webdriver.py", line 59, in __init__
vm-test-run-castopod> client #     self.service.path = DriverFinder.get_path(self.service, options)
vm-test-run-castopod> client #                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vm-test-run-castopod> client #   File "/nix/store/6w72wswxqz79xdg7yzqszqjv0idsa211-python3-3.11.5-env/lib/python3.11/site-packages/selenium/webdriver/common/driver_finder.py", line 41, in get_path
vm-test-run-castopod> client #     raise NoSuchDriverException(msg) from err
vm-test-run-castopod> client # selenium.common.exceptions.NoSuchDriverException: Message: Unable to obtain driver for firefox using Selenium Manager.; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location
```

**Comment by mweinelt** [comment]
I updated #264386 to include a default path for geckodriver again. That makes for example the vaultwarden tests succeed again.

**Comment by AleXoundOS** [comment]
@mweinelt, :pray: the updated https://github.com/NixOS/nixpkgs/pull/264386 fix makes it work!

**Comment by mweinelt** [comment]
Will go via staging-next (cc #263535) and hit master in a few days.

**Comment by AleXoundOS** [comment]
@SuperSandro2000, how do you think, is it ok to generate an mp3 file of size 500 MB in a derivation used in the test? https://github.com/NixOS/nixpkgs/pull/260518/files#diff-a00be17539b40d94bcca7ade61a6e663976485c6a45ac5d76cfb6b843768fe43R34-R46
It uses `sox` and `lame` packages to generate one. But I'm afraid changes in these will provoke rebuilds of the file and take up disk space for Hydra.
It's used only for tests and not part of the castopod package, so it won't go into public cache. And Hydra might garbage collect them?

**Comment by AleXoundOS** [comment]
ZHF: #265948
These fixes are important for castopod and I hope PR gets into NixOS 23.11.

**Comment by sinavir** [comment]
Hello a new version of castopod bas been released. How can I help to update castopod ? ( I already have a custom patch that I wish to upstream)

File: nixos/modules/services/web-apps/castopod.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/castopod.nix:131]:
```suggestion
          https://code.castopod.org/adaures/castopod/-/issues/330).
          
          Note, that practical upload size limit is smaller. For example, with 512 MiB
          setting, around 500 MiB is possible.
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by AleXoundOS** [comment]
@SuperSandro2000, please, review test-time generation of the mp3 file: https://github.com/NixOS/nixpkgs/pull/260518/commits/3c5969133a8b14925ff9465be479eddd110df067.

**Comment by AleXoundOS** [comment]
Hi @sinavir! You can fork nixpkgs and propose changes via a new PR. Or if the changes are small and you don't care about authorship, just post the diff here in comments and I will apply them.

**Comment by AleXoundOS** [comment]
@SuperSandro2000, or maybe instead of https://github.com/NixOS/nixpkgs/pull/260518/commits/3c5969133a8b14925ff9465be479eddd110df067 it's better to use something more complex like https://github.com/NixOS/nixpkgs/blob/33dd4c5bc577ded23b94de0c9417cebb0ae5063e/nixos/lib/utils.nix#L6?

**Comment by sinavir** [comment]
> Hi @sinavir! You can fork nixpkgs and propose changes via a new PR. Or if the changes are small and you don't care about authorship, just post the diff here in comments and I will apply them.

Hello, I finally did a PR, rebasing all your excellent work on master

**Comment by AleXoundOS** [comment]
> > Hi @sinavir! You can fork nixpkgs and propose changes via a new PR. Or if the changes are small and you don't care about authorship, just post the diff here in comments and I will apply them.
> 
> Hello, I finally did a PR, rebasing all your excellent work on master

Great! Please, can you post a link to your commits?

Unfortunately, this PR has halted for a long time, waiting for a review from @SuperSandro2000, @mweinelt or some other nixpkgs maintainer with authority.

**Comment by mweinelt** [comment]
I never planned on reviewing this, I just provided assistance with the selenium issue which I apparently caused.

The question is, whether we want to continue in https://github.com/NixOS/nixpkgs/pull/292835, which contains the 1.10.4 update.

**Comment by sinavir** [comment]
> > > Hi @sinavir! You can fork nixpkgs and propose changes via a new PR. Or if the changes are small and you don't care about authorship, just post the diff here in comments and I will apply them.
> > 
> > 
> > Hello, I finally did a PR, rebasing all your excellent work on master
> 
> Great! Please, can you post a link to your commits?
> 
> Unfortunately, this PR has halted for a long time, waiting for a review from @SuperSandro2000, @mweinelt or some other nixpkgs maintainer with authority.

Hello, the PR is #292835 ! I think first you can review (if you are OK with that ofc) my changes so we agree on the PR content. Then I know some people experienced that proposed to me to review finally the PR. Finally the package is kind of super outdated. I think we have to move on and not wait forever a hypothetical review from @SuperSandro2000

**Comment by sinavir** [comment]
And if you're ok let's move this discussion on #292835 !

**Comment by sinavir** [comment]
Hello, any updates on whether we should continue here or on https://github.com/NixOS/nixpkgs/pull/292835 ?

**Comment by JulienMalka** [comment]
Kindly closing because this PR has been superseded by https://github.com/NixOS/nixpkgs/pull/292835

**Pull request closed without merging**
